### PR TITLE
ARM split - detect USB to select master/slave

### PR DESCRIPTION
**Pull request by zvecr**
## Current Issue
Master/slave detection is hard coded to always assume master

## TL;DR
This PR changes the startup behavior to detect an active USB connection when delegating master/slave. If this operation times out, then the half is assume to be a slave. This is the default behavior for ARM, and required for AVR Teensy boards (due to hardware limitations).

<details><summary>CLICK ME FOR INVESTIGATION INFO</summary>

## master/slave detection
Split code currently expects to run different routines on slave and master. The assumption is that devices cannot run both a requester and responder concurrently. The current process is that the master pulls the matrix and encoder data from the slave, and pushes rgb and backlight settings to the slave.

Current AVR init order:
```
lufa/lufa.c::main
    keyboard.c::keyboard_setup
        split_util.c::matrix_setup
            'usb check'
                init i2c/serial slave/master
    setup_usb
```

Current ARM init order:
```
chibios/main.c::main
    keyboard.c::keyboard_setup
        split_util.c::matrix_setup
            'usb check'
                init i2c/serial slave/master
    init_usb_driver
```

As shown above, 'usb check' cannot simply check "Do I currently have a USB connection". AVR assumes a hardware specific method of checking the usb power. **Note: This check fails on some hardware, as shown here <https://geekhack.org/index.php?topic=89674.0>**

### vitimins_included
Old split code modified to use a slightly different process. Both sides idle as slaves, then one is promoted to master. The main advantage here is that the transport init runs after usb has been configured. 
```
lufa/lufa.c::main
    keyboard.c::keyboard_setup
        vitamins_included::matrix_setup
            "nothing important"
    setup_usb
    keyboard.c::keyboard_init
        vitamins_included::matrix_init
            init i2c/serial slave
            "wait for usb"
                run slave tasks
            init i2c/serial master
```

#### USB connected based
Wrapping up similar behavior to `WAIT_FOR_USB`

* adapt vitimins_included code into split_common
  * Maybe simplify with just a `wait_for_usb_with_timeout`
* init usb earlier
  * allows existing split code to be retained by pushing change elsewhere
* delay usb check check till after setup_usb
    * lazy init of transport to first matrix_scan
    * some processes might assume too much
        * rgblight clips to left and right hand and would need to handle it at a later point
        * pending support for split rgb matrix would need to handle left/right hand changing at a later point

#### Other
Find alternative "usb connected" methods, that allow "meetup RGB demo mode". The keyboard can be connected to a battery pack and thus have no usb connection but still maintain the master/slave connection.

* Using similar method to AVR VBUS detection
  * Fine for custom board where the additional circuitry can be added
  * Bad for Proton C as hardware has already shipped
* Using similar method to SPLIT_HAND_PIN
  * Would allow a `MASTER_LEFT` `MASTER_RIGHT` setup
  * Proton C is limited to bridging pins on the snap off section
    * Less flexible due to hard coded master/slave
    * Provides battery powered "demo mode"

</details>

## Description
Delaying split initialisation from `matrix_setup` to `matrix_init` was seen to be the most non intrusive change. As this has limitations for "battery demo mode", this may not be the full solution but gives us something to use while building up the remaining core split_common functionality.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
## Testing notes
Scenarios currently tested
* Bootmagic full and lite
* RGB light
* Pro micro split
  * lets_split
* Teensy split
  * currently requires `#define NO_USB_STARTUP_CHECK`
* vitamins_included (via @Duckle29)
  * same as Teensy split
  * Potential change instead of `NO_USB_STARTUP_CHECK` is being investigated before vitamins_included is refactored

File: quantum/split_common/split_util.c [hunk 1]
```diff
@@ -15,8 +15,27 @@
...
 #endif
 
+#ifndef SPLIT_USB_TIMEOUT
+#    define SPLIT_USB_TIMEOUT 2500
+#endif
+
 volatile bool isLeftHand = true;
 
+bool waitForUsb(void) {
+    for(uint8_t i = 0; i < (SPLIT_USB_TIMEOUT / 100); i++) {
```
Comment (nooges) [inline quantum/split_common/split_util.c]:
```suggestion
    for (uint8_t i = 0; i < (SPLIT_USB_TIMEOUT / 100); i++) {
```

Comment (zvecr) [reply quantum/split_common/split_util.c]: Ah this is before future, let me run clang-format over the changes.
